### PR TITLE
Refactored and Added AssetBundleVerifierWindow bundle tests

### PR DESCRIPTION
**Pull request by jedouard98**
Refactored code to now support tests for all three error states: 

- DestinationError,
- WebRequestError,
- BundleError

### CODE REVIEW

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 1]
```diff
@@ -25,32 +29,44 @@ public class AssetBundleVerifierWindow : EditorWindow
     {
         private const int FieldMinWidth = 170;
 
-        private bool _assetBundleDownloadIsSuccessful;
-        private string _assetBundleUrl;
+        internal bool AssetBundleDownloadIsSuccessful;
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Group the internal fields together and add comment

// Visible for testing

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 2]
```diff
@@ -25,32 +29,44 @@ public class AssetBundleVerifierWindow : EditorWindow
...
+        internal string ErrorDescription;
+
         private string _mainScene;
         private double _numOfMegabytes;
         private UnityWebRequest _webRequest;
+        private AssetBundle _bundle;
+
+        public AssetBundleVerifyState State;
+
+        internal static string WebRequestErrorFormatMessage = "Problem retrieving AssetBundle from {0}: {1}";
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Make const and move up to top with other const fields

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 3]
```diff
@@ -25,32 +29,44 @@ public class AssetBundleVerifierWindow : EditorWindow
...
+
+        public enum AssetBundleVerifyState
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Make this the first thing in the class, before const fields

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 4]
```diff
@@ -61,41 +77,69 @@ private void StartAssetBundleVerificationDownload()
...
+                    AssetBundleDownloadIsSuccessful = false;
+                    ErrorDescription = "Error extracting AssetBundle. See Console log for details.";
+                    // No need to log since debugging information in this case is automatically logged by Unity.
+                    break;
+                case AssetBundleVerifyState.DownloadSuccess:
+                    AssetBundleDownloadIsSuccessful = true;
+                    _numOfMegabytes = ConvertBytesToMegabytes(webRequest.downloadedBytes);
+                    var scenes = _bundle.GetAllScenePaths();
+                    _mainScene = (scenes.Length == 0) ? "No scenes in AssetBundle" : scenes[0];
+                    break;
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]:
```
default:
  throw new NotImplementedException(string.Format("Unexpected state {0}", state));
```

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 5]
```diff
@@ -61,41 +77,69 @@ private void StartAssetBundleVerificationDownload()
...
+                return AssetBundleVerifyState.InProgress;
+            }
+
+            try
+            {
+                _bundle = DownloadHandlerAssetBundle.GetContent(_webRequest);
+            }
+            catch (InvalidOperationException e)
+            {
+                Debug.LogError(e.ToString());
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Something like this maybe
`Debug.LogErrorFormat("Failed to obtain AssetBundle content: {0}", e);`

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 6]
```diff
@@ -61,41 +77,69 @@ private void StartAssetBundleVerificationDownload()
...
+                return AssetBundleVerifyState.DestinationError;
+            }
 
             _responseCode = _webRequest.responseCode;
 
 #if UNITY_2017_1_OR_NEWER
             if (_webRequest.isHttpError || _webRequest.isNetworkError)
 #else
-            if (_webRequest.isError)
+            if (webRequest.isError)
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Should be _webRequest ?

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 7]
```diff
@@ -25,32 +29,44 @@ public class AssetBundleVerifierWindow : EditorWindow
...
+        public static void ShowWindow()
         {
-            // Set AssetBundle url in a private variable so that information displayed in window is consistent with
-            // the url that this was called on. 
-            var window = (AssetBundleVerifierWindow) GetWindow(typeof(AssetBundleVerifierWindow), true,
-                "Play Instant AssetBundle Verify");
-            window._assetBundleUrl = assetBundleUrl;
-
-            window.StartAssetBundleVerificationDownload();
+            GetWindow(typeof(AssetBundleVerifierWindow), true, "Play Instant AssetBundle Verify");
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: `return GetWindow(typeof(AssetBundleVerifierWindow), true, "Play Instant AssetBundle Verify");`

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 8]
```diff
@@ -25,32 +29,44 @@ public class AssetBundleVerifierWindow : EditorWindow
...
+            WebRequestError,
+            BundleError,
+            DownloadSuccess
+        }
 
         /// <summary>
         /// Creates a dialog box that details the success or failure of an AssetBundle retrieval from a given assetBundleUrl.
         /// </summary>
-        public static void ShowWindow(string assetBundleUrl)
+        public static void ShowWindow()
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: `public static AssetBundleVerifierWindow ShowWindow()`

File: GooglePlayInstant/Editor/QuickDeploy/QuickDeployWindow.cs [hunk 9]
```diff
@@ -218,7 +218,9 @@ private void OnGuiVerifyBundleSelect()
                 }
                 else
                 {
-                    AssetBundleVerifierWindow.ShowWindow(_assetBundleUrl);
+                    var window = (AssetBundleVerifierWindow) GetWindow(typeof(AssetBundleVerifierWindow), true,
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/QuickDeployWindow.cs]: `var window = AssetBundleVerifierWindow.GetWindow();` (see above)

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    /// Contains tests for each of AssetBundleVerifier error states.
+    /// </summary>
+    [TestFixture]
+    public class AssetBundleVerifierTest
+    {
+        [UnityTest]
+        public IEnumerator TestDestinationError()
+        {
+            var window = (AssetBundleVerifierWindow) EditorWindow.GetWindow(typeof(AssetBundleVerifierWindow), true,
+                "Play Instant AssetBundle Verify");
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: `var window = AssetBundleVerifierWindow.GetWindow();` (see above)

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+            while (window == null)
+            {
+                yield return null;
+            }
+
+            window.HandleAssetBundleVerifyState(AssetBundleVerifierWindow.AssetBundleVerifyState.DestinationError,
+                new UnityWebRequest());
+
+            Assert.AreEqual(false, window.AssetBundleDownloadIsSuccessful,
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: `Assert.IsFalse(...`

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+            while (window == null)
+            {
+                yield return null;
+            }
+
+            window.HandleAssetBundleVerifyState(AssetBundleVerifierWindow.AssetBundleVerifyState.WebRequestError,
+                new UnityWebRequest());
+
+            Assert.AreEqual(false, window.AssetBundleDownloadIsSuccessful,
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Same

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+                "AssetBundle download should not be marked successful during destination error state.");
+        }
+
+        [UnityTest]
+        public IEnumerator TestWebRequestError()
+        {
+            var window = (AssetBundleVerifierWindow) EditorWindow.GetWindow(typeof(AssetBundleVerifierWindow), true,
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: same

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+            LogAssert.Expect(LogType.Error,
+                string.Format(AssetBundleVerifierWindow.WebRequestErrorFormatMessage, window.AssetBundleUrl,
+                    window.ErrorDescription));
+        }
+
+        [UnityTest]
+        public IEnumerator TestBundleError()
+        {
+            var window = (AssetBundleVerifierWindow) EditorWindow.GetWindow(typeof(AssetBundleVerifierWindow), true,
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Same

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+            while (window == null)
+            {
+                yield return null;
+            }
+
+            window.HandleAssetBundleVerifyState(AssetBundleVerifierWindow.AssetBundleVerifyState.BundleError,
+                new UnityWebRequest());
+
+            Assert.AreEqual(false, window.AssetBundleDownloadIsSuccessful,
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Same

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    [TestFixture]
+    public class AssetBundleVerifierTest
+    {
+        [UnityTest]
+        public IEnumerator TestDestinationError()
+        {
+            var window = (AssetBundleVerifierWindow) EditorWindow.GetWindow(typeof(AssetBundleVerifierWindow), true,
+                "Play Instant AssetBundle Verify");
+
+            while (window == null)
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: I don't understand why this block is needed:
- Why isn't window immediately set to a non-null value the by the statement above?
- If you remove this while loop does the test not work?
- Can you link to some documentation explaining this?

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 17]
```diff
@@ -61,41 +77,69 @@ private void StartAssetBundleVerificationDownload()
 #endif
         }
 
-        private void GetAssetBundleInfoFromDownload()
+        internal void HandleAssetBundleVerifyState(AssetBundleVerifyState state, UnityWebRequest webRequest)
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Add comment

// Visible for testing

**Review by thanwah — COMMENTED** [review]
Thanks for adding tests!

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 1 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 2 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 3 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 8 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 7 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Needs to be casted, but resolved.

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 17 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 4 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 6 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Sorry, thanks for the catch. Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/QuickDeployWindow.cs (hunk 9 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/QuickDeployWindow.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs (hunk 10 above)
Comment (jedouard98) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Resolved

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs (hunk 11 above)
Comment (jedouard98) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs (hunk 13 above)
Comment (jedouard98) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs (hunk 12 above)
Comment (jedouard98) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs (hunk 14 above)
Comment (jedouard98) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs (hunk 15 above)
Comment (jedouard98) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 5 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs (hunk 16 above)
Comment (jedouard98) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/AssetBundleVerifierTest.cs]: Sorry, I think I miscorrectly diagnosed a problem with the tests to the window object not rendering in time; I was getting NullPointerExceptions previously without the use of this specific block. I did all the changes, and retried it again, but it seems to be working. Not sure what the issue was before. Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs [hunk 18]
```diff
@@ -61,9 +76,57 @@ private void StartAssetBundleVerificationDownload()
 #endif
         }
 
-        private void GetAssetBundleInfoFromDownload()
+        // Visible for testing
+        internal void HandleAssetBundleVerifyState(AssetBundleVerifyState state, UnityWebRequest webRequest)
+        {
+            switch (state)
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: I assume it's intentional that AssetBundleVerifyState.InProgress is not included here?

**Review by thanwah — APPROVED** [review]
LGTM

File: GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs (hunk 18 above)
Comment (jedouard98) [reply GooglePlayInstant/Editor/QuickDeploy/AssetBundleVerifierWindow.cs]: Yeah. That state should never be handled, just added a clarifying comment here.

**Pull request merged**